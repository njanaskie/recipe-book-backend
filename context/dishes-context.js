import React from 'react';

const DishesContext = React.createContext()

export { DishesContext as default }