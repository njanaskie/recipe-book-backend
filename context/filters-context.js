import React, { useContext, useReducer } from 'react';
import filtersReducer, { filtersReducerDefaultState } from '../src/reducers/filters'

const FiltersContext = React.createContext()

export const useFiltersContext = () => useContext(FiltersContext)

const FiltersProvider = ({ children }) => {
    const [filters, filtersDispatch] = useReducer(filtersReducer, filtersReducerDefaultState)

    return (
        <FiltersContext.Provider value={{ filters, filtersDispatch }}>
            {children}
        </FiltersContext.Provider>
    )
}

export { FiltersProvider, FiltersContext as default }