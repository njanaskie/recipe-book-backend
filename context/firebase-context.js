import React, { useEffect, useState, useContext, useReducer } from 'react';
import database, { firebase, googleAuthProvider } from '../src/firebase/firebase';
import { history } from '../src/routers/AppRouter';
import authReducer from '../src/reducers/auth'
import LoadingPage from '../src/components/LoadingPage'

const FirebaseContext = React.createContext()

export const useFirebaseContext = () => useContext(FirebaseContext)

const FirebaseProvider = ({ children }) => {
    const [auth, authDispatch] = useReducer(authReducer, [])
    // const authDispatch = useDispatch()
    const [user, setUser] = useState('')
    const [loading, setLoading] = useState(true)
    const [isAdmin, setIsAdmin] = useState()
    const [isGuest, setIsGuest] = useState()
    const email = process.env.GUEST_EMAIL
    const password = process.env.GUEST_PASSWORD

    const login = () => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(() => {
            authDispatch({ type: 'LOGIN' })
        })
    }

    const loginAsGuest = () => {
        firebase.auth().signInWithEmailAndPassword(email, password).then(() => {
            authDispatch({ type: 'LOGIN_AS_GUEST' })
        })
    }

    const logout = () => {
        firebase.auth().signOut().then(() => {
            authDispatch({ type: 'LOGOUT' })
        });
    }

    useEffect(() => {
        const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
            setUser(user)
            setLoading(false)
            setIsAdmin()
            setIsGuest()
        })

        return () => unsubscribe()
    }, [])

    useEffect(() => {
        if (user) {
            database.collection('users').doc(user.uid).get()
            .then((doc) => {
                const userData = doc.data()
                if (userData) {
                    if (userData.isAdmin === true) {
                        const isAdmin = userData.isAdmin
                        setIsAdmin(isAdmin)
                    } else if (userData.isGuest === true) {
                        const isGuest = userData.isGuest
                        setIsGuest(isGuest)
                    }
                }
            })
        }
    }, [user])

    if (loading) {
        return <LoadingPage />
    }
    // if (user) {
    //     if (history.location.pathname === '/') {
    //         history.push('/dashboard')
    //     }
    // } else {
    //     history.push('/');
    // }

    return (
        <FirebaseContext.Provider value={{ user, isAdmin, isGuest, login, loginAsGuest, logout }}>
            {children}
        </FirebaseContext.Provider>
    )
}

export { FirebaseProvider, FirebaseContext as default }