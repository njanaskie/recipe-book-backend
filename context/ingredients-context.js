import React from 'react';

const IngredientsContext = React.createContext()

export { IngredientsContext as default }