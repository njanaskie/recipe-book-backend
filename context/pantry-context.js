import React, { useContext } from 'react';

const PantryContext = React.createContext()

export const usePantryContext = () => useContext(PantryContext)

export { PantryContext as default }