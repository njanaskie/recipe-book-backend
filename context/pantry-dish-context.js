import React, { useContext } from 'react';

const PantryDishContext = React.createContext()

export const usePantryDishContext = () => useContext(PantryDishContext)

export { PantryDishContext as default }