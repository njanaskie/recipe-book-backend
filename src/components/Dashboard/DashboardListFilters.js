import React, { useState, useEffect, useContext } from 'react'
import { Input, Dropdown } from 'semantic-ui-react'
import FiltersContext from '../../../context/filters-context'
import PantryContext from '../../../context/pantry-context'
import dishCuisines from '../../fixtures/dishCuisines'
import dishTypes from '../../fixtures/dishTypes'

const DashboardListFilters = () => {
    const { filters, filtersDispatch } = useContext(FiltersContext)
    const { pantryIngredients } = useContext(PantryContext)

    const onTextChange = (e) => {
        filtersDispatch({ type: 'SET_TEXT_FILTER', text: e.target.value })
    }

    const onKeyIngredientChange = (e, result) => {
        const { value } = result || e.target
        filtersDispatch({ type: 'SET_KEY_INGREDIENT_FILTER', keyIngredients: value })
    }

    const onCuisineChange = (e, result) => {
        const { value } = result || e.target
        filtersDispatch({ type: 'SET_CUISINE_FILTER', cuisine: value })
    }

    const onTypeChange = (e, result) => {
        const { value } = result || e.target
        filtersDispatch({ type: 'SET_TYPE_FILTER', dishType: value })
    }

    console.log(filters)

    return (
        <div>
            <Input
                placeholder='Search...'
                size='big'
                value={filters ? filters.text : ''}
                onChange={onTextChange}
            />
            <Dropdown
                placeholder='Select key ingredient'
                name='keyIngredient'
                fluid multiple selection
                multiple={true}
                value={filters ? filters.keyIngredients : []}
                onChange={onKeyIngredientChange}
                options={pantryIngredients.map(ingredient => {
                    return {
                        key: ingredient.id,
                        text: ingredient.name,
                        value: ingredient.name
                    }
                })}
            />
            <Dropdown
                placeholder='Select cuisine'
                name='cuisine'
                fluid selection
                value={filters ? filters.cuisine : ''}
                onChange={onCuisineChange}
                options={dishCuisines.map(dishCuisine => {
                    return {
                        key: dishCuisine,
                        text: dishCuisine,
                        value: dishCuisine
                    }
                })}
            />
            <Dropdown
                placeholder='Select type'
                name='type'
                fluid selection
                value={filters ? filters.dishType : ''}
                onChange={onTypeChange}
                options={dishTypes.map(dishType => {
                    return {
                        key: dishType,
                        text: dishType,
                        value: dishType
                    }
                })}
            />
        </div>
    )
}

export default DashboardListFilters