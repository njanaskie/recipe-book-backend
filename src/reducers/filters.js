
// Filters Reducer

const filtersReducerDefaultState = {
    text: '',
    keyIngredients: [],
};

const filtersReducer = (state = filtersReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            }
        case 'SET_KEY_INGREDIENT_FILTER':
            return {
                ...state,
                keyIngredients: action.keyIngredients
            }
        default:
            return state;
        }
    }
    
    export default filtersReducer;