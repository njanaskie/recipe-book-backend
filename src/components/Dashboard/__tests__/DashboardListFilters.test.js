import React from 'react'
import { shallow, mount } from 'enzyme'
import { DashboardListFilters } from '../DashboardListFilters'
import * as AllFiltersContext from '../../../../context/filters-context'
import FiltersContext, { useFiltersContext } from '../../../../context/filters-context'
import { filters } from '../../../tests/fixtures/filters'

let setTextFilter, setKeyIngredientFilter, setCuisineFilter, setTypeFilter, wrapper;

beforeEach(() => {
    setTextFilter = jest.fn()
    setKeyIngredientFilter = jest.fn()
    setCuisineFilter = jest.fn()
    setTypeFilter = jest.fn()
    jest.spyOn(AllFiltersContext, 'useFiltersContext').mockImplementation(() => filters)
    // wrapper = shallow(<DashboardListFilters />, { context: filters }).shallow()
    wrapper = mount(
        <FiltersContext.Provider value={filters}>
            <DashboardListFilters
                filters={filters}
                setTextFilter={setTextFilter}
                setKeyIngredientFilter={setKeyIngredientFilter}
                setCuisineFilter={setCuisineFilter}
                setTypeFilter={setTypeFilter}
            />
        </FiltersContext.Provider>
    )
})

test('should render DashbaordListFilters correctly', () => {
    expect(wrapper).toMatchSnapshot();
})