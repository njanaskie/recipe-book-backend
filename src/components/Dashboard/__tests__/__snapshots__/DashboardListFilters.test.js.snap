// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DashbaordListFilters correctly 1`] = `
<DashboardListFilters
  filters={
    Object {
      "cuisine": "",
      "dishType": "",
      "keyIngredients": Array [],
      "text": "",
    }
  }
  setCuisineFilter={[MockFunction]}
  setKeyIngredientFilter={[MockFunction]}
  setTextFilter={[MockFunction]}
  setTypeFilter={[MockFunction]}
/>
`;
